### ISSUE TITLE
LossBinaryXENT always returns a gradient of 0 if the label is 0.

### ISSUE THREAD

**Issue by bnsh** [deeplearning4j/nd4j#1328]
I think LossBinaryXENT _always_ returns a gradient of 0 *if* the label is 0. Here is a demo that shows that zeros as labels _always_ produce zero gradients: https://gist.github.com/bnsh/a494ecbefa9914d4848930cde17a782c

When I run it, I get:
```
computeGradientAndScore(preout=-5.0000000, label=0.0000000, activation=sigmoid) => {score=0.0067153, gradient=-0.0000000}
computeGradientAndScore(preout=-4.0000000, label=0.0000000, activation=sigmoid) => {score=0.0181500, gradient=-0.0000000}
computeGradientAndScore(preout=-3.0000000, label=0.0000000, activation=sigmoid) => {score=0.0485873, gradient=-0.0000000}
computeGradientAndScore(preout=-2.0000000, label=0.0000000, activation=sigmoid) => {score=0.1269280, gradient=-0.0000000}
computeGradientAndScore(preout=-1.0000000, label=0.0000000, activation=sigmoid) => {score=0.3132617, gradient=-0.0000000}
computeGradientAndScore(preout=0.0000000, label=0.0000000, activation=sigmoid) => {score=0.6931472, gradient=-0.0000000}
computeGradientAndScore(preout=1.0000000, label=0.0000000, activation=sigmoid) => {score=1.3132617, gradient=-0.0000000}
computeGradientAndScore(preout=2.0000000, label=0.0000000, activation=sigmoid) => {score=2.1269281, gradient=-0.0000000}
computeGradientAndScore(preout=3.0000000, label=0.0000000, activation=sigmoid) => {score=3.0485876, gradient=-0.0000000}
computeGradientAndScore(preout=4.0000000, label=0.0000000, activation=sigmoid) => {score=4.0181484, gradient=-0.0000000}
computeGradientAndScore(preout=5.0000000, label=0.0000000, activation=sigmoid) => {score=5.0067191, gradient=-0.0000000}
```

And, here is a demo that computes the finite differences for the gradients as a check, to prove that the gradients are not _truly_ zero. https://gist.github.com/bnsh/d04ce02c2e57c2069196e7263e63a576

When I run this, I get:
```
score(preoutput=3.0000000, label=0.0000000) => 3.0485876 (expect 3.0485874)
Gradient returned by LossBinaryXENT = -0.0000000
xp1=3.0485964 xm1=3.0485773
Gradient computed by finite differences = 0.9536743
```

(xp1 is f(x+epsilon) and xm1 is f(x-epsilon) and I'm computing (f'(x) as f(x+epsilon) - f(x-epsilon)) / (2 * epsilon))

The issue, I think, are these three lines in LossBinaryXENT.java: 69-71

            INDArray outputder = Nd4j.getExecutioner().execAndReturn(Nd4j.getOpFactory().createTransform(activationFn, preOutput.dup()).derivative());
            grad = outputder.muli(labels);
            grad.divi(output).muli(-1);

So, XE(p, l, F) = l * log(F(p)) + (1-l) * log(1-F(p))

We want XE'(p,I,F) wrt p = 
     = (l / F(p) - (1-l) / (1-F(p))) * F'(p)
     = (l(1-F(p)) - (1-l)(F(p))) * F'(p) / (F(p) * (1-F(p))
     = (l - l F(p) - F(p) + l F(p)) * F'(p) / (F(p) * (1-F(p)))
     = (l - F(p)) * F'(p) / (F(P) * (1-F(p)))

(where l=label, p=preoutput, F=activation function (and F(p) = output))

But, those three lines above, instead are doing

    1. outputder = F'(p)
    2. outputder = outputder * labels; grad = outputder
    3. grad = grad / output; grad = grad * -1

Which has the effect of: grad = -outputder * labels / output (or -F'(p) * l / F(p))...

So, of course, grad is 0 if labels == 0.

I do have a fix available, if you can verify that my logic is correct and that the bug I think I have found is not in fact something dumb I have done...

**Comment by bnsh** [comment]
I'm not really familiar with github, but, I think I have a fix here: https://github.com/bnsh/nd4j (specifically: https://github.com/bnsh/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossBinaryXENT.java )

**Comment by eraly** [comment]
I'll take a look and we'll get it squared away by the end of Monday.

**Comment by bnsh** [comment]
Cool.. should I do the "pull request"? Or should I wait till you test it?

**Comment by eraly** [comment]
Let me look at it first.

**Comment by eraly** [comment]
So, a couple of things - 
1. First off, notice how the gradient in the code does not return a double. It returns an ndarray. What is being computed is the gradient wrt to each entry in the preout matrix. So that sum needs to be seen as an ndarray
2. Just intuitively - If one label is zero, changing the preout to that will not affect your loss function (y*log(y_hat) (unless we are talking softmax) so it makes sense that the gradient is zero and should be zero.
3. I am reran the FiniteDifferences.java in the master and I am not seeing what you are seeing. The score is zero (which I am expecting) and the finite differences is also zero (which I am expecting).

So all this kind of boils down to how you are calculating your loss function. You have to have two variables, label and label-1 and check the gradients against preout0 and preout1.

Hope that make sense. I am changing the status of the issue to question. I'll try to attach a picture explaining it if I get a chance later.

**Comment by eraly** [comment]
Also maybe this will help put the pieces in place - with cross entropy you want to have the labels be of minimum size 2 (even if they are l and l-l) otherwise you should pick another loss function and treat it like regression...

**Comment by bnsh** [comment]
I don't think this is the case...

The score for BinaryXENT is simply : -Sum[t_i Log[o_i] + (1-t_i) Log[o_i], {i,1,n}]

Let's have n = 1.  And, let's have o_i = 0.9, and t_i = 0

```
-Sum[t_i Log[o_i] + (1-t_i)Log[(1-o_i)], {i,1,1}] // I made a typo up there.. The second log should have Log[(1-o_i)]
= -(0 Log[0.9] + (1 Log[0.1]))
= -(0 * (-.105) + 1 * (-2.303))
= -(0 - 2.303)
= 2.303
```

Ugh. I should have done it with a pre-out.

Let's have n = 1.  And, let's have p_i = 3, and t_i = 0 and activation=Sigmoid.
o_i = 1/(1+Exp[-p_i]) = 1/(1+Exp[-3]) = .953

```
-Sum[t_i Log[o_i] + (1-t_i)Log[(1-o_i)], {i,1,1}] // I made a typo up there.. The second log should have Log[(1-o_i)]
= -(0 Log[0.953] + (1 Log[0.047]))
= -(0 * (-.048) + 1 * (-3.058))
= -(0 - 3.058)
= 3.058
```

[Cross-entropy error function and logistic regression](https://en.wikipedia.org/wiki/Cross_entropy#Cross-entropy_error_function_and_logistic_regression)

Re: regression vs. classification tho... Despite logistic_regression being called that, I think that's just a historical artifact, and that logistic regression is classifying each input into a single label with two states: 0 and 1.

I'll make another gist in a few minutes to show what I mean with a function that does 2 outputs... Say, "xor" and "and" to illustrate.. (Also, a neural network that will compute that...)

OK, I just went all out instead, and made a neural network that will just compute all 16 possible boolean functions: https://gist.github.com/bnsh/e579912868cf5472ec47a376827dd634

(So, those 16 possible boolean functions are listed in Main.java hopefully that makes sense...)

**Comment by bnsh** [comment]
So, per our gitter conversation tho, I think maybe I'll put down my understanding of the relationship of Softmax with XE to Logistic with XE...

So, there are labels and there are states.

Say, in the case of an ImageNet classifier, each image "has" a label that is in one of a 1000 states. So, it is _either_ a monkey or a steamboat or a rainbow or whatever. The output, is a probability distribution over all the states, and as a probability distribution, they all have to sum to 1.

So, at least in my parlance, there's still only one label, it's just that that label can be in one of a 1000 states (for the target), and the system will output probability values for each of those states, and all those probabilities must sum to 1.

Say if we had another thing that could classify things based on emotional content: say {"happy", "sad", "scary" }, then we'd have a single label, which could be in 3 states.

If we made another network, that then could classify both, so using ComputationGraph, it output both ImageNet labels, and my "emotion" labels, then each image would have 2 labels, and the imagenet one would have 1000 states, and the emotion one would have 3 states.

In this way, we could have the generalized XE function like so:

```
XE = Sum[Sum[Sum[t_ijk Log[o_ijk], {i,1,s_j}], {j, 1,l}],{k,1,m}]
```

Where,
 * m = #cases,
 * l=#labels,
 * s_j=#states for label set j and
 * Sum[t_ixy,{i,1,n}]== 1 and Sum[o_ixy,{i,1,n}] == 1 for any x, y.

So, for our image set case, say, m = # cases (however many images we have), l=2, s_1 = 1000 and s_2 = 3.

In the special case of a *single* BinaryXE tho, we have m = #cases again, l=1, s_1=2

```
Sum[Sum[Sum[t_ijk Log[o_ijk], {i,1,2}], {j, 1,1}],{k,1,m}]
```

But, we can simplify this to

```
Sum[Sum[t_ik Log[o_ik], {i,1,2}],{k,1,m}]
```

We can expand that to
```
= Sum[t_1k Log[o_1k] + t_2k Log[o_2k],{k,1,m}]
```

and knowing that (t_1k + t_2k == 1) for all k and (o_1k + o_2k == 1) for all k we can further simplify that to

```
= Sum[t_1k Log[o_1k] + (1-t_1k) Log[1-o_1k],{k,1,m}]
```

And, if we just assert that t_k is the "state" where whatever is "true", then we can just say

```
= Sum[t_k Log[o_k] + (1-t_k) Log[1-o_k],{k,1,m}]
```

And, that's what I'm doing in MyLossBinaryXENT... (And, it agrees with Torch's [BCECriterion](https://github.com/torch/nn/blob/master/doc/criterion.md#nn.BCECriterion) (minus the size averaging bit that you'll see there in the form of -1/n, and adjusting their "i" for my "k"...)

**Comment by eraly** [comment]
@AlexDBlack Can you take a look? 
Long and short of it is: Currently for binary classification we are expecting the user to have two output nodes (even if they sum up to one). The assertion from the user is that our binary cross entropy function should be rewritten to only take one output. And the gradient should reflect that. I am fine with what we have now as it gives us a certain amount of flexibility,

**Comment by bnsh** [comment]
"One output"... is a misunderstanding. As I show in my Main.java gist, there, the cross entropy function as I've written it can take multiple outputs. It's just that each is a separate "bit" as it were...  They don't have to sum to 1. So, my example has for each pair of binary inputs 16 outputs, each of which computes all the 16 possible binary functions.

**Comment by eraly** [comment]
So with our current setup if you wanted to use binary cross entropy and build 16 different binary functions we would need 16 different nets. Whereas if we assumed the output is represented as the user describes you could in theory build it all in one go. So I guess this comes down to what is considered more common. Although I like the concept of keeping these separate so that they can be used to predict probabilities for independent classes.
And it is true that our binary cross entropy is the same as multi class entropy for classes=2.

**Comment by AlexDBlack** [comment]
Binary xent should work for both single output and multiple output cases in the same implementation.
In the multiple output case, nOut(x) means x binary classes, which may be present or absent independently (i.e., unlike softmax which assumes mutually exclusive classes). x=1 gives us binary classification; x=2 gives us 2 binary classes, and so on.

The loss for each output is given by https://en.wikipedia.org/wiki/Cross_entropy#Cross-entropy_error_function_and_logistic_regression
Then we take the sum of the losses for the x output classes.

**Comment by bnsh** [comment]
So.. It's wrong that even for a single output, that the gradient should _always_ be zero, right? My issue is not a non-issue?

**Comment by AlexDBlack** [comment]
Confirming this as an issue. DL4J loss function gradient checks for this are failing.

```java
    public static void main(String[] args){
        for( int i=-5; i<=5; i++ ) {

            LossBinaryXENT loss = new LossBinaryXENT();
            INDArray preOut = Nd4j.create(new double[]{i});
            INDArray labels0 = Nd4j.create(new double[]{0});
            INDArray labels1 = Nd4j.create(new double[]{1});

            INDArray grad0 = loss.computeGradient(labels0, preOut, "sigmoid", null);
            double score0 = loss.computeScore(labels0, preOut, "sigmoid", null, true);
            INDArray grad1 = loss.computeGradient(labels1, preOut, "sigmoid", null);
            double score1 = loss.computeScore(labels1, preOut, "sigmoid", null, true);

            System.out.println(i + "\tScores:\t" + score0 + "\t" + score1 + "\tGradients:\t" + grad0.getDouble(0) + "\t" + grad1.getDouble(0));
        }
    }
```
Scores look reasonable there, but gradients are wrong for the label = 0 case. There should be a symmetry there in the gradients, like the scores.

**Comment by bnsh** [comment]
Thanks @AlexDBlack.. I think I have a fix at https://github.com/bnsh/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossBinaryXENT.java .. If you want to verify it..

**Comment by AlexDBlack** [comment]
I had a quick lok at your code. Output/gradients look sensible on my test above, and it passes DL4J gradient checks.
ScoreArray math looks good to me.

**Comment by bnsh** [comment]
Awesome... Should I make a pull request?

**Comment by AlexDBlack** [comment]
Yep, that'd be great.

**Comment by AlexDBlack** [comment]
Fixed: https://github.com/deeplearning4j/nd4j/pull/1338
